sdks/go/pkg/beam/core/runtime/pipelinex/replace.go

```diff
@@ -243,3 +245,33 @@ func findFreeName(seen map[string]bool, name string) string {
 		}
 	}
 }
+
+// ApplySdkImageOverrides takes a pipeline and a map of patterns to overrides,
+// and proceeds to replace matching ContainerImages in any Environments
+// present in the pipeline.
+func ApplySdkImageOverrides(p *pipepb.Pipeline, patterns map[string]string) error {
+	if len(patterns) == 0 {
+		return nil
+	}
+
+	for _, env := range p.GetComponents().GetEnvironments() {
+		var payload pipepb.DockerPayload
+		if err := proto.Unmarshal(env.GetPayload(), &payload); err != nil {
+			return err
+		}
+		oldImg := payload.GetContainerImage()
+		for pattern, replacement := range patterns {
+			re, err := regexp.Compile(pattern)
```

Comment:
No action required, just a call out to be wary of nested for loops and repeated work.

I think my main concern here is that we're compiling the patterns multiple times (each once per environment at most), which is probably not going to be terrible since this is at construction time, and both patterns and environments are going to be limited.

Reply:
Good point. I went ahead and fixed it anyway, and yeah I'll keep that in mind.